c++/src/capnp/rpc.c++

```diff
@@ -2766,6 +2774,9 @@ private:
         tasks.add(KJ_EXCEPTION(DISCONNECTED, "Peer disconnected."));
         return false;
       }
+    }, [this](kj::Exception&& exception) {
+      receiveIncomingMessageError = true;
+      return kj::mv(exception);
```

Comment:
I think this is coercing the exception value to `bool` rather than propagating it, because the first callback passed to `then()` returns `bool` -- the second callback's return value always gets coerced to the same type as the first callback's. (Though, I'm surprised that doesn't lead to a compiler error, exceptions really shouldn't be convertible to `bool`.)

Try this instead:

```
kj::throwRecoverableException(kj::mv(exception));
return false;
```